util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -74,195 +93,217 @@ var (
+	informer, err := getInformer(w.getCustomResourceLabelSelector(), namespace, k8sClient, gvr)
+	if err != nil {
+		klog.Fatalf("Error getting informer:%s", err)
 	}
-	w.startListening(workerListenPort, handlers)
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			w.handleCustomResource(obj)
+		},
+	})
+	w.stopCh = make(chan struct{})
+	informer.Run(w.stopCh)
 }
 
-func (w *Worker) startListening(port string, handlers handlersMap) {
-	for urlPath, handler := range handlers {
-		http.HandleFunc(urlPath, handler)
-	}
-	if err := http.ListenAndServe(":"+port, nil); err != nil {
-		klog.Fatalf("Failed to start http server on port %v: %v", port, err)
+func (w *Worker) populatePodName() {
+	var isPodNameAvailable bool
+	w.podName, isPodNameAvailable = os.LookupEnv("POD_NAME")
+	if !isPodNameAvailable {
+		klog.Fatal(errors.New("pod name not set as environment variable"))
 	}
+	klog.Info("Pod Name set:", w.podName)
 }
 
-// Metrics returns the metrics collected.
-func (w *Worker) Metrics(rw http.ResponseWriter, request *http.Request) {
-	select {
-	case <-w.stopCh:
-		if w.err != nil {
-			message := fmt.Sprintf("metrics collection failed: %v", w.err)
-			w.sendResponse(rw, http.StatusInternalServerError, Status{Message: message})
-			return
-		}
-		reply := MetricResponse{
-			Result:      w.parsedResult,
-			WorkerDelay: w.workerDelay,
-		}
-		w.sendResponse(rw, http.StatusOK, reply)
-	default:
-		w.sendResponse(rw, http.StatusInternalServerError, Status{Message: "metric collection in progress"})
-	}
+func (w *Worker) getCustomResourceLabelSelector() string {
+	return fmt.Sprintf("%s in (clientPodName,serverPodName)", w.podName)
 }
 
-func (w *Worker) sendResponse(rw http.ResponseWriter, statusCode int, response interface{}) {
-	rw.Header().Set("Content-Type", "application/json")
-	marshalledResponse, err := json.Marshal(response)
+func (w *Worker) handleCustomResource(obj interface{}) {
+	newRuntimeObj, ok := obj.(runtime.Object)
+	if obj != nil && !ok {
+		klog.Errorf("Error casting object: %s", obj)
+		return
+	}
+	err := w.populateResourceSpec(newRuntimeObj)
 	if err != nil {
-		klog.Errorf("Error marshalling to json: %v", err)
-		rw.WriteHeader(http.StatusInternalServerError)
+		err = fmt.Errorf("populating resource spec failed: %v", err)
+		w.updateStatus(makeErrorStatus(err))
+		klog.Error(err)
 		return
 	}
-	klog.V(3).Infof("Marshalled Response: %v", response)
-	rw.WriteHeader(statusCode)
-	if _, err := rw.Write(marshalledResponse); err != nil {
-		klog.Errorf("Error writing response to ResponseWriter: %v", err)
+	klog.Info("Recevied add event for resource with spec:", w.work.resourceSpec)
+	switch w.podName {
+	case w.work.resourceSpec.ClientPodName:
+		w.work.workType = "client"
+	case w.work.resourceSpec.ServerPodName:
+		w.work.workType = "server"
+	default:
+		klog.Error(errors.New("pod name not set as client or server"))
+		w.updateStatus(makeErrorStatus(err))
```

Comment:
you're reporting the wrong error here - see also comment above

Reply:
Updated.